src/cursor.rs

```diff
@@ -21,48 +37,63 @@ use crate::{error::Result, operation::GetMore, options::StreamAddress, Client, N
 /// results from the server; both of these factors should be taken into account when choosing the
 /// optimal batch size.
 ///
-/// A cursor can be used like any other [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html). The simplest way is just to iterate over the
+/// A cursor can be used like any other [`Stream`](https://docs.rs/futures/0.3.4/futures/stream/trait.Stream.html). The simplest way is just to iterate over the
 /// documents it yields:
 ///
 /// ```rust
+/// # use futures::stream::StreamExt;
 /// # use mongodb::{Client, error::Result};
 /// #
-/// # fn do_stuff() -> Result<()> {
+/// # async fn do_stuff() -> Result<()> {
 /// # let client = Client::with_uri_str("mongodb://example.com")?;
 /// # let coll = client.database("foo").collection("bar");
-/// # let cursor = coll.find(None, None)?;
+/// # let mut cursor = coll.find(None, None)?;
 /// #
-/// for doc in cursor {
+/// while let Some(doc) = cursor.next().await {
 ///   println!("{}", doc?)
 /// }
 /// #
 /// # Ok(())
 /// # }
 /// ```
 ///
-/// Additionally, all the other methods that an [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html) has are available on `Cursor` as well.
+/// Additionally, all the other methods that an [`Stream`](https://docs.rs/futures/0.3.4/futures/stream/trait.Stream.html) has are available on `Cursor` as well.
+/// This includes all of the functionality provided by [`StreamExt`](https://docs.rs/futures/0.3.4/futures/stream/trait.StreamExt.html), which provides similar functionality to the standard library `Iterator` trait.
 /// For instance, if the number of results from a query is known to be small, it might make sense
 /// to collect them into a vector:
 ///
 /// ```rust
 /// # use bson::{doc, bson, Document};
+/// # use futures::stream::StreamExt;
 /// # use mongodb::{Client, error::Result};
 /// #
-/// # fn do_stuff() -> Result<()> {
+/// # async fn do_stuff() -> Result<()> {
 /// # let client = Client::with_uri_str("mongodb://example.com")?;
 /// # let coll = client.database("foo").collection("bar");
 /// # let cursor = coll.find(Some(doc! { "x": 1 }), None)?;
 /// #
-/// let results: Vec<Result<Document>> = cursor.collect();
+/// let results: Vec<Result<Document>> = cursor.collect().await;
 /// # Ok(())
 /// # }
 /// ```
 #[derive(Debug)]
 pub struct Cursor {
     client: Client,
-    buffer: VecDeque<Document>,
     get_more: GetMore,
     exhausted: bool,
+
+    /// This will be None when the cursor is exhausted.
+    state: Option<PollState>,
```

Comment:
Rather than storing this as an `Option<PollState>`, I think it would be more straightforward if we just introduced a third state called `Exhausted`. This would remove the need to document what `None` means here and below. Given that `Exhausted` wouldn't necessarily be a state of polling, it may make sense to rename it to `CursorState` or something.

Reply:
I've updated it as you suggested. Since the enum is private to the cursor module, I just named it `State`. I also needed to move things around in the `poll_next` implementation a bit; because we're not using an option anymore, we can't match on `self.state.take()`, so I had to switch all of the captures to take references and then resolve all of the borrow checker errors that resulted.